axum/src/body/mod.rs

```diff
@@ -22,12 +23,31 @@ pub use bytes::Bytes;
 pub type BoxBody = http_body::combinators::UnsyncBoxBody<Bytes, Error>;
 
 /// Convert a [`http_body::Body`] into a [`BoxBody`].
+///
+/// If the given body is already a [`BoxBody`], this function will not box the body again.
 pub fn boxed<B>(body: B) -> BoxBody
 where
     B: http_body::Body<Data = Bytes> + Send + 'static,
     B::Error: Into<BoxError>,
 {
-    body.map_err(Error::new).boxed_unsync()
+    downcast_box_body(body).unwrap_or_else(|body| body.map_err(Error::new).boxed_unsync())
+}
+
+fn downcast_box_body<B: 'static>(body: B) -> Result<BoxBody, B> {
+    let mut body = Some(body);
+    if let Some(body) = <dyn Any>::downcast_mut::<Option<BoxBody>>(&mut body) {
+        Ok(body.take().unwrap())
+    } else {
+        Err(body.unwrap())
+    }
+}
```

Comment:
We already have [this](https://github.com/tokio-rs/axum/blob/main/axum/src/routing/mod.rs#L660-L672) which does the same thing. Maybe we could move that to the root module and re-use it?